src/name_utils.rs

```diff
@@ -0,0 +1,253 @@
+const ROLE_NAME_MIN_LEN: u8 = 1;
+const ROLE_NAME_MAX_LEN: u8 = 10;
+const GROUP_NAME_MIN_LEN: u8 = 1;
+const GROUP_NAME_MAX_LEN: u8 = 8;
+const NODE_NAME_MIN_LEN: u8 = 0;
+const NODE_NAME_MAX_LEN: u8 = 10;
+const MISC_MIN_LEN: u8 = 0;
+const MISC_MAX_LEN: u8 = 5;
+
+/// The sub names that make up the full resource name.
+#[derive(Debug, strum_macros::Display, strum_macros::EnumString, EnumIter)]
+enum SubName {
+    /// CustomResourceDefinition short name
+    Short,
+    /// The CustomResource name
+    Cluster,
+    /// The CustomResource role name
+    Role,
+    /// The CustomResource role group name
+    Group,
+    /// The optional node name (e.g. for pods)
+    Node,
+    /// Miscellaneous identifiers (e.g. "data" or "conf" for config maps)
+    Misc,
+}
+
+impl SubName {
+    /// Returns the minimum length for each sub name
+    fn min(&self) -> u8 {
+        match self {
+            SubName::Short => SHORT_NAME_MIN_LEN,
+            SubName::Cluster => CLUSTER_NAME_MIN_LEN,
+            SubName::Role => ROLE_NAME_MIN_LEN,
+            SubName::Group => GROUP_NAME_MIN_LEN,
+            SubName::Node => NODE_NAME_MIN_LEN,
+            SubName::Misc => MISC_MIN_LEN,
+        }
+    }
+
+    /// Returns the maximum length for each sub name
+    fn max(&self) -> u8 {
+        match self {
+            SubName::Short => SHORT_NAME_MAX_LEN,
+            SubName::Cluster => CLUSTER_NAME_MAX_LEN,
+            SubName::Role => ROLE_NAME_MAX_LEN,
+            SubName::Group => GROUP_NAME_MAX_LEN,
+            SubName::Node => NODE_NAME_MAX_LEN,
+            SubName::Misc => MISC_MAX_LEN,
+        }
+    }
+}
+
+/// Build a Kubernetes resource name. This is intended to work with the generatedName (add a unique
+/// hash) from Kubernetes. This method ensures that all single components do not exceed a certain
+/// length in order to keep the resource name below 63 (the maximum allowed characters) minus an
+/// offset for the unique hash.
+///
+/// In each sub name (`short_name`, `cluster_name`...) all non alphanumeric parts are automatically
+/// removed. After the removal each sub_name is cut off if it exceeds its specified length.
+///
+/// After processing each sub name is concatenated with "-" and "-" as the last character (to
+/// separate from the kubernetes hash).
+///
+/// # Arguments
+///
+/// * `short_name` - The short name of the custom resource definition.
+/// * `cluster_name` - The name of the custom resource.
+/// * `role_name` - The role name of the custom resource.
+/// * `group_name` - The group name of the custom resource.
+/// * `node_name` - Optional node name if available (e.g. for pods).
+/// * `misc_name` - Optional miscellaneous identifiers (e.g. "data" for config maps).
+///
+pub fn build_resource_name(
+    short_name: &str,
+    cluster_name: &str,
+    role_name: &str,
+    group_name: &str,
+    node_name: Option<&str>,
+    misc_name: Option<&str>,
+) -> OperatorResult<String> {
+    let mut full_name = String::new();
```

Comment:
If `full_name` were a vec you could just append to it and call `join("-")` at the end.